File: lint-arwaky/src-python/infrastructure/stdio_transport_client.py
"""Stdio transport — direct subprocess execution, no DesktopCommander needed."""

from __future__ import annotations

import asyncio
import os
from pathlib import Path

from ..contract import ICommandExecutorPort
from ..taxonomy import FilePath, ResponseData, Timeout, PatternList


class StdioClient(ICommandExecutorPort):
    """Direct subprocess transport. Runs commands locally without any intermediary."""

    def __init__(self, timeout: Timeout = Timeout(value=300.0)):
        self.timeout = timeout.value

    async def execute_command(
        self,
        command: PatternList,
        working_dir: FilePath = FilePath(value="."),
        timeout: Timeout | None = None,
    ) -> ResponseData:
        """Execute command via subprocess and capture output."""
        resolved_dir = str(Path(working_dir.value).resolve())
        timeout_val = timeout.value if timeout else self.timeout
        cmd_list = command.values

        try:
            proc = await asyncio.create_subprocess_exec(
                *cmd_list,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                cwd=resolved_dir,
                env={**os.environ, "PYTHONUNBUFFERED": "1"},
            )

            try:
                stdout_bytes, stderr_bytes = await asyncio.wait_for(
                    proc.communicate(), timeout=timeout_val
                )
            except asyncio.TimeoutError:
                proc.kill()
                await proc.wait()
                return ResponseData(
                    stdout="",
                    stderr=f"Command timed out after {timeout_val}s",
                    returncode=-1,
                    metadata={
                        "protocol": "Stdio",
                        "error": "timeout",
                    },
                )

            return ResponseData(
                stdout=stdout_bytes.decode("utf-8", errors="replace"),
                stderr=stderr_bytes.decode("utf-8", errors="replace"),
                returncode=proc.returncode if proc.returncode is not None else -1,
                metadata={"protocol": "Stdio"},
            )

        except FileNotFoundError:
            return ResponseData(
                stdout="",
                stderr=f"Command not found: {cmd_list[0]}",
                returncode=127,
                metadata={"protocol": "Stdio", "error": "command_not_found"},
            )
        except Exception as e:
            return ResponseData(
                stdout="",
                stderr=str(e),
                returncode=1,
                metadata={"protocol": "Stdio", "error": str(e)},
            )

    async def health_check(self) -> ResponseData:
        """Stdio is always healthy if Python is running."""
        import time

        start = time.time()
        result = await self.execute_command(
            PatternList(values=["echo", "health"]), FilePath(value=".")
        )
        latency_ms = (time.time() - start) * 1000

        if result.returncode == 0:
            return ResponseData(
                stdout=result.stdout,
                stderr=result.stderr,
                returncode=0,
                metadata={
                    "status": "healthy",
                    "latency_ms": round(latency_ms, 2),
                    "protocol": "Stdio",
                    "pid": os.getpid(),
                },
            )
        return ResponseData(
            stdout=result.stdout,
            stderr=result.stderr,
            returncode=result.returncode,
            metadata={
                "status": "unhealthy",
                "error": result.stderr or "Unknown",
                "protocol": "Stdio",
            },
        )

    def close(self):
        """No resources to clean up for stdio."""
        pass

File: lint-arwaky/src-rust/infrastructure/stdio_transport_client.rs
/// stdio_transport_client — Direct subprocess execution transport.
use crate::contract::ICommandExecutorPort;
use crate::taxonomy::*;
use std::time::Duration;
use tokio::process::Command;
use async_trait::async_trait;
use std::collections::HashMap;

pub struct StdioClient {
    timeout: Duration,
}

impl StdioClient {
    pub fn new(timeout: Duration) -> Self {
        Self { timeout }
    }
}

#[async_trait]
impl ICommandExecutorPort for StdioClient {
    async fn execute_command(&self, command: PatternList, working_dir: FilePath, timeout: Option<Duration>) -> anyhow::Result<ResponseData> {
        let timeout_val = timeout.unwrap_or(self.timeout);
        let cmd_list: Vec<&str> = command.values.iter().map(|s| s.as_str()).collect();
        if cmd_list.is_empty() {
            anyhow::bail!("Empty command");
        }
        let mut cmd = Command::new(cmd_list[0]);
        if cmd_list.len() > 1 {
            cmd.args(&cmd_list[1..]);
        }
        cmd.current_dir(&working_dir.value).env("PYTHONUNBUFFERED", "1");
        cmd.kill_on_drop(true);

        let result = tokio::time::timeout(timeout_val, cmd.output()).await;
        match result {
            Ok(Ok(output)) => {
                let mut meta_map = HashMap::new();
                meta_map.insert("protocol".to_string(), serde_json::Value::String("Stdio".to_string()));
                
                Ok(ResponseData {
                    value: serde_json::Value::Null,
                    stdout: StdOutput::new(String::from_utf8_lossy(&output.stdout).to_string()),
                    stderr: StdError::new(String::from_utf8_lossy(&output.stderr).to_string()),
                    returncode: ExitCode::new(output.status.code().unwrap_or(-1) as i64),
                    metadata: MetadataVO::new(meta_map),
                })
            }
            Ok(Err(e)) => anyhow::bail!("Command execution failed: {}", e),
            Err(_) => {
                anyhow::bail!("Command timed out after {}s", timeout_val.as_secs())
            }
        }
    }

    async fn health_check(&self) -> anyhow::Result<ResponseData> {
        Ok(ResponseData::new(serde_json::json!({"status": "ok"})))
    }
}
